### PR TITLE
Add ring buffers mapped to userspace

### PR DESCRIPTION
**Pull request by ai-tmpst**

### CODE REVIEW

File: fw/ringbuffer.c
Comment (const-t) [inline fw/ringbuffer.c:89]: Why do you use mem barrier with per-cpu data?

Comment (ai-tmpst) [reply fw/ringbuffer.c:89]: You are right, it always executes in softirq context, and preemption disabling is excessive. Will get rid of it.

File: fw/ringbuffer.c
Comment (EvgeniiMekhanik) [inline fw/ringbuffer.c:54]: We don't need `preempt_disable` as we discussed in private discussion since this function is executed in softirq

File: fw/ringbuffer.c
Comment (EvgeniiMekhanik) [inline fw/ringbuffer.c:43]: I am not sure but it seems that we can use just
```
int
tfw_ringbuffer_write(TfwStr *str)
```
because TfwStr can contains a lot of TfwStr as chunks.

File: fw/ringbuffer.c
Comment (EvgeniiMekhanik) [inline fw/ringbuffer.c:59]: If we pass single TfwStr with a lot of chunks, we can skip this loop.

File: fw/ringbuffer.c
Comment (EvgeniiMekhanik) [inline fw/ringbuffer.c:148]: It seems that this function is used only for test purpose, maybe move it to unit tests?

Comment (ai-tmpst) [reply fw/ringbuffer.c:148]: In this function we change a field of an internal object.
OK, maybe I can move this field to TfwRingbufer structure.

File: fw/ringbuffer.c
Comment (krizhanovsky) [inline fw/ringbuffer.c:191]: We don't keep testing-specific code in the main code, designated for the real workload. I'd move the function to `test_ringbuffer.c` and just include `ringbuffer.c` to `test_ringbuffer.c`

File: fw/ringbuffer.c
Comment (krizhanovsky) [reply fw/ringbuffer.c:89]: I think the point is that atomics imply full barriers on x86-64

File: fw/ringbuffer.h
Comment (krizhanovsky) [inline fw/ringbuffer.h:64]: It makes sense to express the constants using `PAGE_SIZE`

**Review by krizhanovsky — APPROVED** [review]
LGTM after cleanups mentioned by @const-t and @EvgeniiMekhanik . But please also have a look onto other RB implementations - maybe we can get ready to use implementation. For now the implementation lacks the file operations, but later we might face a bigger problem to let user-space threads sleep on no events and make it low overhead

File: fw/ringbuffer.h
Comment (krizhanovsky) [inline fw/ringbuffer.h:75]: Could you please align the data structure members and comment them, in both the new structures. Not everything is clear from a quick look

File: fw/ringbuffer.c
Comment (krizhanovsky) [inline fw/ringbuffer.c:145]: So we're going to have per-cpu file, right? Could you please comment the function about this consideration? Basically having a minimal documentation in the source, like I referened in the linux kernel, would be useful

File: fw/ringbuffer.c
Comment (krizhanovsky) [inline fw/ringbuffer.c:248]: What happens if we unload the module while the user space still didn't close and/or unmap the file?

**Review by krizhanovsky — COMMENTED** [review]
There are some logic absent - please address it in a next PR

File: fw/ringbuffer.c
Comment (krizhanovsky) [inline fw/ringbuffer.c:1]: Technically this is the second lock-free RB in Tempesta FW and it's not generic. I'd rename the file to something pointing out that this is a kernel/user-space transport, not just an abstract data RB

File: fw/ringbuffer.c
Comment (ai-tmpst) [reply fw/ringbuffer.c:89]: Oops, I answered to the comment about preempt_disable.

So we can do without smp_mb, don't we?

File: fw/ringbuffer.h
Comment (ai-tmpst) [reply fw/ringbuffer.h:26]: Actually my current version also require data length knowledge before the call. 
Maybe we get rid of it if we allow incomplete data blocks. As you mentioned early, we could send fields as TLV and mark the end of the data block with a special "end" TLV. So we could get the size of the room between tail and head before writing, and on every field addition we have to make sure if there is enough place for this field and ending TLV.

Will add the motivation to the comments in the code.

File: fw/ringbuffer.h
Comment (ai-tmpst) [reply fw/ringbuffer.h:64]: I wanted to define this for userspace too.
OK, will find another way to do it.

File: fw/ringbuffer.c
Comment (ai-tmpst) [reply fw/ringbuffer.c:145]: No, just one file with multiple mapped per-cpu areas.

There is no appropriate munmap field in `struct file_operations`. I could set vma->vm_ops and catch unmap event, but I think it would be better to move this field to the Ringbuffer structure, so a client could change it when it prepared to read.

File: fw/ringbuffer.c
Comment (ai-tmpst) [reply fw/ringbuffer.c:248]: The memory remains available to user space.
I thought about it and wanted to notify the user space with a special event. But maybe it would be more correctly to handle this situation on the RB level. Will think.

File: fw/ringbuffer.c
Comment (const-t) [reply fw/ringbuffer.c:89]: > So we can do without smp_mb, don't we?

Yes, we don't need `smp_mb`

File: fw/ringbuffer.h
Comment (krizhanovsky) [reply fw/ringbuffer.h:64]:
```
$ getconf -a|grep 4096
PATH_MAX                           4096
_POSIX_PATH_MAX                    4096
PIPE_BUF                           4096
_POSIX_PIPE_BUF                    4096
NL_ARGMAX                          4096
PAGESIZE                           4096
PAGE_SIZE                          4096
POSIX_ALLOC_SIZE_MIN               4096
POSIX_REC_MIN_XFER_SIZE            4096
POSIX_REC_XFER_ALIGN               4096
```

File: fw/ringbuffer.c
Comment (krizhanovsky) [reply fw/ringbuffer.c:43]: We agreed on the call that this API introduces unwanted overhead on forming an array, so probably we should just provide a raw memory and a commit call, which moves the `tail`

File: fw/t/unit/test_mmap_buffer.c
Comment (krizhanovsky) [inline fw/t/unit/test_mmap_buffer.c:124]: The API looks good and I didn't get from https://github.com/tempesta-tech/tempesta/pull/2259#discussion_r1799440292

> Actually my current version also require data length knowledge before the call.

It seems now we get as much memory as RB has and can write as much data as we need (and can), so there is no need to compute the size of all written data upfront.

File: fw/mmap_buffer.c
Comment (krizhanovsky) [inline fw/mmap_buffer.c:105]: too wide line: we that's OK to slightly exceed 80 characters per line, but not so much

File: fw/mmap_buffer.c
Comment (krizhanovsky) [inline fw/mmap_buffer.c:255]: `holders_cnt` can exceed `MAX_HOLDERS`

File: fw/mmap_buffer.h
Comment (krizhanovsky) [inline fw/mmap_buffer.h:149]: Coding style: could you please align the data structure members and adjust comments like for example in https://github.com/tempesta-tech/tempesta/blob/master/fw/cache.c#L129 ?

File: fw/mmap_buffer.c
Comment (krizhanovsky) [inline fw/mmap_buffer.c:215]: It seems you need to zeroize the memory (or GFP_ZERO) since you check `holder->pg[cpu] ` for `NULL` in `tfw_mmap_buffer_free()`

File: fw/mmap_buffer.c
Comment (krizhanovsky) [inline fw/mmap_buffer.c:180]: I din't get the sense of `(void)dev_vm_ops` statement...

File: fw/mmap_buffer.c
Comment (krizhanovsky) [inline fw/mmap_buffer.c:190]:
```suggestion
static void
dev_file_vm_close(struct vm_area_struct *vma)
```

**Review by krizhanovsky — APPROVED** [review]
I commented some issues, but there are also pending reviews from @const-t and @EvgeniiMekhanik , so I just approve since there is no any design issues, and leave the rest for your fixes and reviews.

File: fw/mmap_buffer.c
Comment (ai-tmpst) [reply fw/mmap_buffer.c:72]: The problem is - we allocate the buffer by `alloc_pages_node(cpu_to_node(cpu), GFP_KERNEL, order)`, and there is two bad way: we use division or we lose almost a half of the buffer size (minus header size).
Maybe we can use another function for allocation which doesn't restrict memory size by a power of 2. But I haven't found appropriate one.
I don't know... Maybe we can use __alloc_percpu and get struct page pointers walking through every allocated page. I considered this way but it looked a little bit ugly for me. Is there another ways?

File: fw/t/unit/test_mmap_buffer.c
Comment (ai-tmpst) [reply fw/t/unit/test_mmap_buffer.c:124]: Yes, in this variant we don't need to compute whole event size before writing.

File: fw/mmap_buffer.c
Comment (ai-tmpst) [reply fw/mmap_buffer.c:215]: Yes, I saw it and replaced by kzalloc.

File: fw/mmap_buffer.c
Comment (ai-tmpst) [reply fw/mmap_buffer.c:180]: Didn't notice, left after debugging.

**Comment by krizhanovsky** [comment]
Should we just close it in favor of https://github.com/tempesta-tech/tempesta/pull/2273 ?

File: fw/mmap_buffer.c
Comment (EvgeniiMekhanik) [inline fw/mmap_buffer.c:214]: Broken aligment

File: fw/mmap_buffer.c
Comment (EvgeniiMekhanik) [inline fw/mmap_buffer.c:206]: Broken aligment

File: fw/mmap_buffer.c
Comment (EvgeniiMekhanik) [inline fw/mmap_buffer.c]: broken aligment

File: fw/mmap_buffer.c
Comment (EvgeniiMekhanik) [inline fw/mmap_buffer.c]: It seems that `__alloc_percpu_gfp` can return NULL, so we should handle this case

File: fw/mmap_buffer.c
Comment (EvgeniiMekhanik) [inline fw/mmap_buffer.c:251]: class_create can fails, so we should check dev_class

File: fw/mmap_buffer.c
Comment (EvgeniiMekhanik) [inline fw/mmap_buffer.c:252]: device_create can also fails.so we should handle this case

File: fw/t/unit/test_mmap_buffer.c
Comment (EvgeniiMekhanik) [inline fw/t/unit/test_mmap_buffer.c:180]: We should add simple test to check device createion.
I add
```
TEST(tfw_mmap_buffer, create_dev)
{
	holder = tfw_mmap_buffer_create("test", TFW_MMAP_BUFFER_MIN_SIZE);
	EXPECT_NOT_NULL(holder);
	EXPECT_NULL(tfw_mmap_buffer_create("test", TFW_MMAP_BUFFER_MIN_SIZE));
	tfw_mmap_buffer_free(holder);
}
```
And kernel crashes because we don't handle invalid device creation

File: fw/mmap_buffer.c
Comment (EvgeniiMekhanik) [inline fw/mmap_buffer.c:297]: It seems that deinitialization should be more complicated, because device can be NULL and class can be NULL

File: fw/mmap_buffer.c
Comment (EvgeniiMekhanik) [inline fw/mmap_buffer.c:222]: `alloc_percpu_gfp` can fails we should handle this case

**Review by EvgeniiMekhanik — CHANGES_REQUESTED** [review]
Need fixes

File: fw/mmap_buffer.c
Comment (ai-tmpst) [reply fw/mmap_buffer.c:59]: DONE

File: fw/mmap_buffer.c
Comment (ai-tmpst) [reply fw/mmap_buffer.c:206]: I used 4-spaces tabs. Fixed.

File: fw/mmap_buffer.c
Comment (ai-tmpst) [reply fw/mmap_buffer.c:214]: I used 4-spaces tabs. Fixed.

File: fw/mmap_buffer.c
Comment (ai-tmpst) [reply fw/mmap_buffer.c:222]: Agree, will fix.

File: fw/mmap_buffer.c
Comment (ai-tmpst) [reply fw/mmap_buffer.c:251]: Fixed.

File: fw/mmap_buffer.c
Comment (ai-tmpst) [reply fw/mmap_buffer.c:252]: Fixed.

File: fw/mmap_buffer.c
Comment (ai-tmpst) [reply fw/mmap_buffer.c:277]: Fixed.

File: fw/mmap_buffer.c
Comment (ai-tmpst) [reply fw/mmap_buffer.c:297]: Fixed.

File: fw/mmap_buffer.h
Comment (ai-tmpst) [reply fw/mmap_buffer.h:149]: Fixed.

File: fw/t/unit/test_mmap_buffer.c
Comment (ai-tmpst) [reply fw/t/unit/test_mmap_buffer.c:180]: Done.

File: fw/mmap_buffer.c
Comment (ai-tmpst) [reply fw/mmap_buffer.c:105]: Fixed.

File: fw/mmap_buffer.c
Comment (ai-tmpst) [reply fw/mmap_buffer.c:190]: Fixed.

File: fw/ringbuffer.c
Comment (ai-tmpst) [reply fw/ringbuffer.c]: Fixed.

**Comment by ai-tmpst** [comment]
Looks through all the comments. All issues are fixed.
This PR can be closed. Let's continue work in PR [2273](https://github.com/tempesta-tech/tempesta/pull/2273).

**Pull request closed without merging**
